### PR TITLE
AWS: abort S3 input stream on close if not EOS

### PR DESCRIPTION
**Pull request by bryanck**
This PR adds a check when closing an S3 input stream if the stream is at end-of-stream. If not, it calls `abort()` on the stream instead of `close()`. This avoids reading to the end of the stream when closing.

The Apache HTTP connection will always read the [entire stream](https://github.com/apache/httpcomponents-core/blob/2135596b1319910d56efb148566ecbbec0884638/httpcore/src/main/java/org/apache/http/impl/io/ContentLengthInputStream.java#L108) on close, which results in much more data being read than needed. The URL connection client does not behave this way.

Now that the Apache HTTP client is the default instead of URL connection client for AWS, this change should prevent a performance regression.

### CODE REVIEW

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 1]
```diff
@@ -194,9 +195,27 @@ private void openStream() throws IOException {
...
+  private void closeStream() {
     if (stream != null) {
-      stream.close();
+      // if we aren't at the end of the stream, and the stream is abortable, then
+      // call abort() so we don't read the remaining data
+      abortStream();
+      try {
+        stream.close();
+      } catch (Exception e) {
+        // close quietly
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: We might want to log the underlying reason as info or warn here, just so we don't silently swallow issues that may come up.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 2]
```diff
@@ -212,9 +212,7 @@ public void write(byte[] b, int off, int len) throws IOException {
   }
 
   private void newStream() throws IOException {
-    if (stream != null) {
-      stream.close();
-    }
+    closeStream();
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:215]: This change seems unrelated and I'm not sure what it really adds.  It looks like we're dereferencing the stream, but we still would throw in any situation where it actually have any affect.

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:215]: This will always throw on an aborted stream on a checksum check failure, so it could be very noisy.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java (hunk 1 above)
Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: This will always throw on an aborted stream on a checksum check failure, so it could be very noisy

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 2 above)
Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:215]: `stream.close()` was being called without the null check in one place, so this was added to prevent a potential NPE, however this is unrelated to the performance regression.

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java]: I rolled this back

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java (hunk 1 above)
Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java]: I added a trace log

**Comment by singhpk234** [comment]
Thanks @bryanck, this is great find and fix !

one doubt will it cause issue when we switch back the client from apache to url-connection (let's say via table props) considering this was not a issue with url-connection client ? 

Also do we need this in 1.2.1 considering 1.2.0 was still shipped with url-connection as default http client ? The regression commit is still in master [commit](https://github.com/apache/iceberg/commit/053028ceb6d3bc04c223a521600b053235bf344a).

**Comment by bryanck** [comment]
@singhpk234 The `abort()` is a no-op with the URL connection client, so it should behave mostly the same as before when using that client (except with the added read for the EOS check).

I thought we wanted to get https://github.com/apache/iceberg/pull/7119 into 1.2.1, is that still the case @danielcweeks @nastra ? Without that, it becomes more cumbersome to configure, as including the URL connection client on the classpath then requires the default AWS http client be set also (e.g. via system properties).

**Comment by singhpk234** [comment]
@bryanck apologies, as per my understanding, I thought it's other way round, based on this file [UrlConnectionHttpClient.java](https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java#L432-L434) `abort()` actually disconnects the connection and in the close it's a no-op [close()](https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java#L143-L145)

I am also very curious about this statement from the [doc](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/ResponseInputStream.html) as well :
 
"Input stream that provides access to the unmarshalled POJO response returned by the service in addition to the streamed contents. This input stream should be closed to release the underlying connection back to the connection pool.

If it is not desired to read remaining data from the stream, you can explicitly abort the connection via [abort()](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/ResponseInputStream.html#abort()). Note that this will close the underlying connection and require establishing an HTTP connection which may outweigh the cost of reading the additional data."

If we go by the last paragraph and IIUC, it says calling abort will not let the connection being re-used and the connection pool will have to establish a new http connection, can it cause regression in some scenario ?

I am just trying to clear my understanding here. 


-----

> I thought we wanted to get https://github.com/apache/iceberg/pull/7119 into 1.2.1, is that still the case

My understanding was 1.2.1 was only for bug fixes for 1.2.0 release.

**Comment by bryanck** [comment]
@singhpk234 `abort()` is being called on the stream, in this case it is a `ResponseInputStream`, rather than on the client. You are correct that calling `abort()` will render the connection not reusable, so there is a trade off. I tried to balance that by checking for EOS and only aborting if we aren't at the EOS, so in cases we are we can reuse the connection. The trade off is reading a bunch of data unnecessarily vs reusing the connection.

**Comment by bryanck** [comment]
I ran the TPC-DS benchmark with and without the changes in this PR using the Apache HTTP client. Without this change, the result was ~2x slower (as a result of reading far more data). With this change, the result was the same as with URL connection client (and the amount of data transferred was also the same).

**Review by nastra — APPROVED** [review]

**Comment by bryanck** [comment]
@singhpk234 you are right that changing the default HTTP client for a patch release is not appropriate, so we'd leave that out of 1.2.1. We'd like the Apache HTTP client fixes in for 1.2.1, though, so that can be used.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 3]
```diff
@@ -194,9 +195,29 @@ private void openStream() throws IOException {
...
+        // check exception with the Apache HTTP client
+        LOG.trace("Error closing stream", e);
+      }
+      stream = null;
+    }
+  }
+
+  private void abortStream() {
+    try {
+      if (stream instanceof Abortable && stream.read() != -1) {
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:216]: why do we want to read one more byte here? It might cause one more request. I think it does not hurt to even abort when it's already fully read.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 4]
```diff
@@ -194,9 +195,29 @@ private void openStream() throws IOException {
...
-      stream.close();
+      // if we aren't at the end of the stream, and the stream is abortable, then
+      // call abort() so we don't read the remaining data with the Apache HTTP client
+      abortStream();
+      try {
+        stream.close();
+      } catch (Exception e) {
+        // log at trace level as closing an aborted stream will throw a content length
+        // check exception with the Apache HTTP client
+        LOG.trace("Error closing stream", e);
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: That means we will see exceptions for every stream close if trace is enabled. If it is expected to fail with the content length check, can we at least skip that? Then we don't need to log in trace, but we can log in maybe a warning level.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java (hunk 3 above)
Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:216]: If you abort the connection then it will be invalidated and removed from the pool, so it won't be reused. So this is an optimization to attempt to reuse the connection in cases where it has been fully read.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java (hunk 4 above)
Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: @singhpk234 @danielcweeks are you ok if I revert this back to a no-op?

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: I guess the question to ask before this is, why do we need to close it again if it's already aborted? If you see in the Trino logic I referenced, it's an if-else.

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: That would work for the Apache HTTP client, the close doesn't do anything for that. But aborting the stream opened by URL connection client doesn't do anything, so we need to close it in that case.

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: I see, that's interesting difference, thanks for the explanation!

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 5]
```diff
@@ -194,9 +195,29 @@ private void openStream() throws IOException {
...
+    }
+  }
+
+  private void abortStream() {
+    try {
+      if (stream instanceof Abortable && stream.read() != -1) {
+        ((Abortable) stream).abort();
+      }
+    } catch (Exception e) {
+      LOG.debug("An error occurred while aborting the stream", e);
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:220]: Is this expected? If not we should log with warn

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java]: This shouldn't happen, I went ahead and changed this to `warn`. I also went ahead and removed the `trace` log for `close()`.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 6]
```diff
@@ -194,9 +195,28 @@ private void openStream() throws IOException {
...
+  private void closeStream() {
     if (stream != null) {
-      stream.close();
+      // if we aren't at the end of the stream, and the stream is abortable, then
+      // call abort() so we don't read the remaining data with the Apache HTTP client
+      abortStream();
+      try {
+        stream.close();
+      } catch (Exception e) {
+        // close quietly, closing an aborted stream will throw a content length
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: Sorry I guess I did not express it clearly. What I meant was that, can we check for the exception that if it is a content length check exception then we can skip, otherwise we log a warning?

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: I see. `ContentLengthInputStream` is an Apache HTTP class, so referring to that will tie S3InputStream to that client.

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: I could check the name of the class, but that seems like a little bit of a hack.

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: Ah okay, what does the exception look like? I though there are some error code that we can get out of the exception that would imply this specific error. Is that not there?

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: [This](https://github.com/apache/httpcomponents-core/blob/2135596b1319910d56efb148566ecbbec0884638/httpcore/src/main/java/org/apache/http/impl/io/ContentLengthInputStream.java#L147)  is what gets thrown. Calling close on the aborted stream will attempt a read which then throws that.

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: Checking for the class name `ConnectionClosedException` might be the best option we have...? We can throw other IOExceptions like before, and skip if the name is that.

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:206]: Checking `ConnectionClosedException` seems good enough to me, because it makes sense that if the connection is already closed then we can omit the error.

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java]: I think `org.apache.http.ConnectionClosedException` is already in the class path given we have some REST client integration in core, but I might be wrong.

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java]: Sounds good, I pushed that change

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java]: That is a different version than being used by the AWS SDK (v5 instead of v4). Also, the AWS bundle shades the library.

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java]: Another option is we check which HTTP client is being used and change the close behavior based on that.

**Review by jackye1995 — APPROVED** [review]
Thanks, looks good to me!

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 7]
```diff
@@ -196,7 +197,29 @@ private void openStream() throws IOException {
...
-      stream.close();
+      // if we aren't at the end of the stream, and the stream is abortable, then
+      // call abort() so we don't read the remaining data with the Apache HTTP client
+      abortStream();
+      try {
+        stream.close();
+      } catch (IOException e) {
+        // the Apache HTTP client will throw a ConnectionClosedException
+        // when closing an aborted stream, which is expected
+        if (!e.getClass().getSimpleName().equals("ConnectionClosedException")) {
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: Can we just do `instanceof` here?

Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: Oh okay I see this thread https://github.com/apache/iceberg/pull/7262#discussion_r1156495423, since it's specific to Apache Http client, it's not guaranteed it'll be on the classpath so we can't use instanceof

Comment (bryanck) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:208]: That would add a dependency on the Apache HTTP client. Also, the AWS bundle shades the library, so the classes will be different depending on that (different packages).

**Review by amogh-jahagirdar — APPROVED** [review]
LGTM, thanks for the fix @bryanck !

**Comment by jackye1995** [comment]
> Another option is we check which HTTP client is being used and change the close behavior based on that.


Actually that's a potentially better approach, since we only have 2 officially supported client types `apache` and `urlconnection`, and their behaviors are quite different based on this exploration. 

I am a bit worried about the 1 additional byte read, seems to only benefit `urlconnection` but could be safe or worse for `apache`. Switch behavior by client type would solve the issue.

We could get the value from `awsProperties`, but one issue is that for users with a custom client factory, they might overwrote the entire client factory and thus have a HTTP client type mismatch. But even in that case, user can explicitly set the HTTP client to change the behavior, so it seems to be a safe approach.

@bryanck @amogh-jahagirdar any thoughts?

**Comment by bryanck** [comment]
I was hoping we could get it from S3Client but I didn't see a way, I'll look some more tonight

**Comment by bryanck** [comment]
I don't see a good way to get that, without more extensive changes, though I'd be happy to be proven wrong.

**Comment by bryanck** [comment]
The only way I could find to determine the HTTP client type on the S3 client is to use reflection. So we could do that which is not ideal. Another options is to read the AwsProperties to get the type, but that may not match the actual HTTP client for custom factories. I feel like the current solution is the least invasive for the 1.2.1 release and achieves the desired outcome of performance parity between the two HTTP clients.

There are alternatives to detecting the EOS also. We could get the content length of the response and track the number of bytes read from the stream, but I feel like that is heavier than the one byte read. Generaly that byte will be buffered if not EOS so shouldn't trigger any network I/O.

**Review by danielcweeks — APPROVED** [review]

**Review by singhpk234 — COMMENTED** [review]
LGTM as well (apologies being late to party), many thanks @bryanck for this fix !
